transport/owr_crypto_utils.c

```diff
@@ -0,0 +1,230 @@
+  X509 *cert;
+
+  X509_NAME *name = NULL;
+
+
+  EVP_PKEY *key_pair;
+ 
+  RSA *rsa;
+
+  #define GST_DTLS_BIO_BUFFER_SIZE 4096
+  BIO *bio_cert;
+  gchar buffer_cert[GST_DTLS_BIO_BUFFER_SIZE] = { 0 };
+  gint len_cert;
+  gchar *pem_cert = NULL;
+  BIO *bio_key;
+  gchar buffer_key[GST_DTLS_BIO_BUFFER_SIZE] = { 0 };
+  gint len_key;
+  gchar *pem_key = NULL;
+
+  gboolean errorDetected = FALSE;
+
+  bio_cert = BIO_new (BIO_s_mem ());
+  bio_key = BIO_new (BIO_s_mem ());
+
+  GString *string_fprint = NULL;
+  gchar *char_fprint = NULL;
+  guint j;
+  const EVP_MD *fprint_type = NULL;
+  fprint_type = EVP_sha256();
+  guchar fprint[EVP_MAX_MD_SIZE];
+
+  guint fprint_size;
+
+  cert = X509_new ();
+
+  key_pair = EVP_PKEY_new ();
+
+  rsa = RSA_generate_key (2048, RSA_F4, NULL, NULL);
+
+  EVP_PKEY_assign_RSA (key_pair, rsa);
+
+  X509_set_version (cert, 2);
+  ASN1_INTEGER_set (X509_get_serialNumber (cert), 0);
+  X509_gmtime_adj (X509_get_notBefore (cert), 0);
+  X509_gmtime_adj (X509_get_notAfter (cert), 31536000L);  /* A year */
+  X509_set_pubkey (cert, key_pair);
+
+  name = X509_get_subject_name (cert);
+  X509_NAME_add_entry_by_txt (name, "C", MBSTRING_ASC, (unsigned char *) "SE",
+      -1, -1, 0);
+  X509_NAME_add_entry_by_txt (name, "CN", MBSTRING_ASC,
+      (unsigned char *) "OpenWebRTC", -1, -1, 0);
+  X509_set_issuer_name (cert, name);
+  name = NULL;
+
+  X509_sign (cert, key_pair, EVP_sha256 ());
+
+  if (!X509_digest(cert, fprint_type, fprint, &fprint_size)) {
+    g_print("Error creating the certificate fingerprint.\n");
+    errorDetected = TRUE;
+  }
+
+  string_fprint = g_string_new (NULL);
+
+  for (j=0; j < fprint_size; j++) { 
+    g_string_append_printf(string_fprint, "%02X%c", fprint[j], (j+1 == fprint_size) ?'\n':':'); 
+  }
+
+  char_fprint = g_string_free(string_fprint, FALSE);
+
+  if (!PEM_write_bio_X509 (bio_cert, (X509 *) cert)) {
+    g_print("could not write certificate bio \n");
+    errorDetected = TRUE;
+  }
+
+
+  if (!PEM_write_bio_PrivateKey (bio_key, (EVP_PKEY *) key_pair, NULL, NULL, 0, 0, NULL)) {
+    g_print("could not write PrivateKey bio \n");
+    errorDetected = TRUE;
+  }
+
+  len_cert = BIO_read (bio_cert, buffer_cert, GST_DTLS_BIO_BUFFER_SIZE);
+  if (!len_cert) {
+    g_print("no cert length \n");
+    errorDetected = TRUE;
+  }
+
+  len_key = BIO_read (bio_key, buffer_key, GST_DTLS_BIO_BUFFER_SIZE);
+  if (!len_key) {
+    g_print("no key length \n");
+    errorDetected = TRUE;
+  }
+
+  pem_cert = g_strndup (buffer_cert, len_cert);
+  pem_key = g_strndup (buffer_key, len_key);
+
+
+    GValue params[3] = { G_VALUE_INIT, G_VALUE_INIT, G_VALUE_INIT };
+
+    g_value_init(&params[0], G_TYPE_STRING);
+    g_value_init(&params[1], G_TYPE_STRING);
+    g_value_set_string(&params[1], pem_cert);
+    g_value_init(&params[2], G_TYPE_STRING);
+    g_value_set_string(&params[2], char_fprint);
+
+  if (errorDetected) {
+    g_print("error detected \n");
+    g_value_set_string(&params[0], "Failure");
```

Comment:
Perhaps errors should be handled by a 4th argument, and all of these should be set to null. That would open up for giving some more descriptive reasons in the future. We haven't really standardised to a specific error forwarding scheme though, so perhaps we should just keep this as is and maybe change it in the future.